jni/src/dartjni.c

```diff
@@ -368,6 +368,122 @@ JniResult newObject(jclass cls, jmethodID ctor, jvalue* args) {
   return jniResult;
 }
 
+JniPointerResult newPrimitiveArray(jsize length, int type) {
+  attach_thread();
+  void* pointer;
+  switch (type) {
+    case booleanType:
+      pointer = (*jniEnv)->NewBooleanArray(jniEnv, length);
+      break;
+    case byteType:
+      pointer = (*jniEnv)->NewByteArray(jniEnv, length);
+      break;
+    case shortType:
+      pointer = (*jniEnv)->NewShortArray(jniEnv, length);
+      break;
+    case charType:
+      pointer = (*jniEnv)->NewCharArray(jniEnv, length);
+      break;
+    case intType:
+      pointer = (*jniEnv)->NewIntArray(jniEnv, length);
+      break;
+    case longType:
+      pointer = (*jniEnv)->NewLongArray(jniEnv, length);
+      break;
+    case floatType:
+      pointer = (*jniEnv)->NewFloatArray(jniEnv, length);
+      break;
+    case doubleType:
+      pointer = (*jniEnv)->NewDoubleArray(jniEnv, length);
+      break;
+    case objectType:
+    case voidType:
+      // This error should have been handled in dart.
+      // is there a way to mark this as unreachable?
+      // or throw exception in Dart using Dart's C API.
+      break;
+  }
+  JniPointerResult result = {.id = to_global_ref(pointer), .exception = NULL};
+  result.exception = check_exception();
+  return result;
+}
+
+JniPointerResult newObjectArray(jsize length,
+                                jclass elementClass,
+                                jobject initialElement) {
+  attach_thread();
+  jarray array = to_global_ref(
+      (*jniEnv)->NewObjectArray(jniEnv, length, elementClass, initialElement));
+  JniPointerResult result = {.id = array, .exception = NULL};
+  result.exception = check_exception();
+  return result;
+}
+
+JniResult getArrayElement(jarray array, int index, int type) {
+  JniResult result = {NULL, NULL};
+  attach_thread();
+  // Boundary checks
+  if (index < 0 || index >= (*jniEnv)->GetArrayLength(jniEnv, array)) {
+    JniClassLookupResult classRes =
+        getClass("java/lang/ArrayIndexOutOfBoundsException");
+    if (classRes.exception) {
+      result.exception = classRes.exception;
+      return result;
+    }
+    JniPointerResult ctorRes = getMethodID(classRes.classRef, "<init>", "()V");
```

Comment:
I didn't get that. Can you elaborate?

Reply:
> Index 4 out of bounds for length 4

I don't see that information passed anywhere.

Reply:
I see. The problem is `Get<Primitive>ArrayElements` gives a pointer to the first element, so `ptr[index]` would SEGV instead of throwing an exception.

Reply:
Changing this to use `ArrayRegion` instead might fix it and in that case I won't have to create my own exception.